lib/unparser/emitter.rb

```diff
@@ -496,5 +520,41 @@ def run(emitter, node = node())
       emitter.new(node, self).write_to_buffer
     end
 
+    # For binary or unary operators, check whether parentheses are needed
+    # around a nested operator expression.
+    #
+    # @return [Boolean]
+    #
+    # @api private
+    #
+    def needs_parentheses?(emitter, side)
+      # TODO: Later this can also be used for things like if/unless/until/while
+      # statement modifiers
+      # Right now it is just for nested operator expressions
+      if operator? && emitter.operator?
+        parent_prec = OPERATOR_PRECEDENCE[operator]
+        child_prec  = OPERATOR_PRECEDENCE[emitter.operator]
+        parent_intuit = INTUITIVE_PRECEDENCE[operator]
+        child_intuit  = INTUITIVE_PRECEDENCE[emitter.operator]
+        if parent_prec && child_prec && parent_intuit && child_intuit
+          # if omitting parens would change the meaning, we definitely need them
+          return true if child_prec < parent_prec
+          # if it is obvious to an "average" Ruby programmer that parens are not
+          # needed, then we can omit them
+          return false if child_intuit > parent_intuit
+          # in chains of identical operators, like a + b + c + d, we can omit
+          # parens if associativity allows us to do so without changing meaning
+          # + and - are a special case; a + b - c + d is OK without parens
+          return true if operator != emitter.operator &&
+                         (!INTUITIVE_PRECEDENCE_EQUAL.include?(operator) ||
+                         !INTUITIVE_PRECEDENCE_EQUAL.include?(emitter.operator))
+          associativity = OPERATOR_ASSOCIATIVITY[operator]
+          return false if associativity == side
+        end
+      end
+
+      !emitter.terminated?
```

Comment:
> The fact we cannot break this big method up, without adding the smell to pass emitter and side around all the time, might also highlight a missing object.

That's a good point. The method is actually pretty big with or without short-circuiting, so refactoring it would be a priority.